cpp/tests/types/dispatcher_test.cu

```diff
@@ -13,122 +13,160 @@
-    case GDF_DATE64:
-      return (std::is_same<T, cudf::date64>::value);
-    default:
-      return (false);
-    }
-  }
+  /**---------------------------------------------------------------------------*
+   * @brief Lists every gdf_dtype that the type_dispatcher supports.
+   *
+   * This vector *must* list every gdf_dtype supported by the type_dispatcher.
+   *
+   * If a new type gdf_dtype is added, but this list is not updated, then the
+   * tests will fail.
+   *
+   *---------------------------------------------------------------------------**/
+  std::vector<gdf_dtype> supported_dtypes{
+      GDF_INT8,    GDF_INT16,  GDF_INT32,  GDF_INT64,     GDF_FLOAT32,
+      GDF_FLOAT64, GDF_DATE32, GDF_DATE64, GDF_TIMESTAMP, GDF_CATEGORY};
+
+  // These types are not supported by the type_dispatcher
+  std::vector<gdf_dtype> unsupported_dtypes{GDF_invalid, GDF_STRING};
 };
 
-__global__ 
-void dispatch_test_kernel(gdf_dtype type, bool * d_result)
-{
-  if(0 == threadIdx.x + blockIdx.x * blockDim.x)
-    *d_result = cudf::type_dispatcher(type, test_functor{}, type);
+TEST_F(DispatcherTest, NumberOfTypesTest) {
+  // N_GDF_TYPES indicates how many enums there are in `gdf_dtype`,
+  // therefore, if a gdf_dtype is added without updating this test, the test
+  // will fail
+  const size_t expected_num_supported_dtypes =
+      N_GDF_TYPES - unsupported_dtypes.size();
+
+  // Note: If this test fails, that means a type was added to gdf_dtype
+  // without adding it to the `supported_dtypes` list in this test fixture
+  ASSERT_EQ(expected_num_supported_dtypes, supported_dtypes.size())
+      << "Number of supported types does not match what was expected.";
+
+  ASSERT_EQ(expected_num_supported_dtypes,
+            std::tuple_size<decltype(supported_types)>::value);
 }
 
-} // anonymous namespace
+namespace {
+template <typename ExpectedType>
+struct type_tester {
+  template <typename DispatchedType>
+  bool operator()() {
+    return std::is_same<ExpectedType, DispatchedType>::value;
+  }
+};
+}  // namespace
+
+// Ensure that the type_to_gdf_dtype trait maps to the correct gdf_dtype
+TEST_F(DispatcherTest, TraitsTest) {
+  cudf::detail::for_each(supported_types, [](auto type_dummy) {
+    using T = decltype(type_dummy);
+    EXPECT_TRUE(
+        cudf::type_dispatcher(cudf::gdf_dtype_of<T>(), type_tester<T>{}));
+  });
+}
 
-struct DispatcherTest : public GdfTest 
-{
-  std::vector<gdf_dtype> dtype_enums{
-    GDF_INT8, GDF_INT16, GDF_INT32, GDF_INT64, GDF_FLOAT32, 
-    GDF_FLOAT64, GDF_DATE32, GDF_DATE64, GDF_TIMESTAMP, GDF_CATEGORY};
+namespace {
+struct test_functor {
+  template <typename T>
+  __host__ __device__ bool operator()(gdf_dtype type_id) {
+    return (type_id == cudf::gdf_dtype_of<T>());
+  }
 };
 
-TEST_F(DispatcherTest, HostDispatchFunctor)
-{
-  for (auto const &t : this->dtype_enums) {
+__global__ void dispatch_test_kernel(gdf_dtype type, bool* d_result) {
+  if (0 == threadIdx.x + blockIdx.x * blockDim.x)
```

Comment:
I think this line can be removed. Even if you launch multiple threads, it's OK for them all to run the next line.

Reply:
```suggestion
```